src/box/box.cc

```diff
@@ -1900,59 +1913,80 @@ box_wait_limbo_acked(double timeout)
 
 /** Write and process a PROMOTE request. */
 static int
-box_issue_promote(uint32_t prev_leader_id, int64_t promote_lsn)
+box_issue_promote(int64_t promote_lsn)
 {
+	int rc = 0;
 	struct raft *raft = box_raft();
 	uint64_t term = raft->term;
-	assert(raft->volatile_term == term);
+	uint64_t promote_term = txn_limbo.promote_greatest_term;
 	assert(promote_lsn >= 0);
 
+	if (raft->volatile_term != term) {
+		diag_set(ClientError, ER_INTERFERING_ELECTIONS);
```

Comment:
Since you factored this check out, you can use it here as a function too. Same in `box_issue_demote()`.
```Diff
@@ -1916,28 +1916,20 @@ static int
 box_issue_promote(int64_t promote_lsn)
 {
 	int rc = 0;
-	struct raft *raft = box_raft();
-	uint64_t term = raft->term;
+	uint64_t term = box_raft()->term;
 	uint64_t promote_term = txn_limbo.promote_greatest_term;
 	assert(promote_lsn >= 0);
-
-	if (raft->volatile_term != term) {
-		diag_set(ClientError, ER_INTERFERING_ELECTIONS);
-		return -1;
-	}
-
+	rc = box_check_election_term_intact(term);
+	if (rc != 0)
+		goto end;
 	txn_limbo_begin(&txn_limbo);
-
 	rc = box_check_election_term_intact(term);
 	if (rc != 0)
 		goto end;
-
 	rc = box_check_promote_term_intact(promote_term);
 	if (rc != 0)
 		goto end;
-
 	rc = txn_limbo_write_promote(&txn_limbo, promote_lsn, term);
-
 end:
 	if (rc == 0) {
 		txn_limbo_commit(&txn_limbo);
```
Up to you.

Reply:
Thanks for noticing! Cant jump to end there, because it will call `txn_limbo_rollback` without `txn_limbo_begin`, this fails assertion in `latch_unlock()`: `l->owner == fiber()' failed`.
<details>

<summary>
This is the final diff:
</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 70b1a7ebf..970425264 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -1916,26 +1916,20 @@ static int
 box_issue_promote(int64_t promote_lsn)
 {
        int rc = 0;
-       struct raft *raft = box_raft();
-       uint64_t term = raft->term;
+       uint64_t term = box_raft()->term;
        uint64_t promote_term = txn_limbo.promote_greatest_term;
        assert(promote_lsn >= 0);
-
-       if (raft->volatile_term != term) {
-               diag_set(ClientError, ER_INTERFERING_ELECTIONS);
-               return -1;
-       }
+       rc = box_check_election_term_intact(term);
+       if (rc != 0)
+               return rc;

        txn_limbo_begin(&txn_limbo);
-
        rc = box_check_election_term_intact(term);
        if (rc != 0)
                goto end;
-
        rc = box_check_promote_term_intact(promote_term);
        if (rc != 0)
                goto end;
-
        rc = txn_limbo_write_promote(&txn_limbo, promote_lsn, term);

 end:
@@ -1945,7 +1939,6 @@ end:
        } else {
                txn_limbo_rollback(&txn_limbo);
        }
-
        return rc;
 }

@@ -1957,26 +1950,21 @@ static int
 box_issue_demote(int64_t promote_lsn)
 {
        int rc = 0;
-       struct raft *raft = box_raft();
-       uint64_t term = raft->term;
+       uint64_t term = box_raft()->term;
        uint64_t promote_term = txn_limbo.promote_greatest_term;
        assert(promote_lsn >= 0);

-       if (raft->volatile_term != term) {
-               diag_set(ClientError, ER_INTERFERING_ELECTIONS);
-               return -1;
-       }
+       rc = box_check_election_term_intact(term);
+       if (rc != 0)
+               return rc;

        txn_limbo_begin(&txn_limbo);
-
        rc = box_check_election_term_intact(term);
        if (rc != 0)
                goto end;
-
        rc = box_check_promote_term_intact(promote_term);
        if (rc != 0)
                goto end;
-
        rc = txn_limbo_write_demote(&txn_limbo, promote_lsn, term);

 end:
```

</details>